contributors/design-proposals/network/service-topology-key.md

```diff
@@ -0,0 +1,171 @@
+* Performance benefit (node local/rack local) is lower latency/higher bandwidth.
+
+## Background
+
+It has been a very often requested feature because people need to create a service which connects to local backend Pods, the so called local service. In this way, they can reduce network latency, improve security, save money and so on. However, because topology is arbitrary, zone, region, rack, generator, whatever, who knows? We should allow arbitrary locality.
+
+`ExternalTrafficPolicy` was added in v1.4, but only for NodePort and external LB traffic. NodeName was added to `EndpointAddress` to allow kube-proxy to filter local endpoints for various future purposes.
+
+Based on our experience of advanced routing setup and recent demo of enabling this in Kubernetes service, this document would like to introduce a more generic way to enable local service.
+
+## Proposal
+
+This proposal builds off of earlier requests to [use local pods only for kube-proxy loadbalancing](https://github.com/kubernetes/kubernetes/issues/7433) and [node-local service proposal](https://github.com/kubernetes/kubernetes/pull/28637). But, this document proposes that not only the particular "node-local" user case should be taken care, but also a more generic way should be figured out.
+
+Locality is an "user-defined" thing. When we set topology key "hostname" for service, we expect node carries different node labels on the key "hostname".
+
+Users can control the level of topology. For example, if someone run logging agent as a daemonset, he can set the "hard" topology requirement for same-host. If "hard" is not met, then just return "service not available". 
+
+And if someone set a "soft" topology requirement for same-host, say he "preferred" same-host endpoints and can accept other hosts when for some reasons local service's backend is not available on some host.
+
+If multiple endpoints satisfy the "hard" or "soft" topology requirement, we will randomly pick one by default. Routing decision is expected to be implemented in L3/4 VIP level such as kube proxy.
+
+## Implementation details
+
+### API changes
+
+#### Service API changes
+
+As with scheduling topology, there are currently two types of topology, called `Required` and `Preferred` which denote “hard” vs. “soft” requirements. 
+
+Similarly, we introduce a topology mode and an associated topology key in service topology.
+
+```go
+type Service struct {
+  Spec:
+    // topology is used to achieve "local" service in a given topology level.
+    // User can control what ever topology level they want.
+    // +optional
+    Topology ServiceTopology `json:"topology" protobuf:"bytes,1,opt,name=topology"`
+}
+
+// Defines a service topolgoy information.
+type ServiceTopology struct {
+    // Valid values for mode are "ignored", "required", "preferred".
+    // "ignored" is the default value and the associated topology key will have no effect.
```

Comment:
Do we really need "ignored"?  Seems like just leaving the key blank would be the same thing and less work/code.